### PR TITLE
Wait until migration job completed

### PR DESCRIPTION
**Pull request by tomahawk28**
## Checklist

- [ ] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: service/controller/app/resource/releasemigration/create.go [hunk 1]
```diff
@@ -115,6 +115,21 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	// It means helm release migration is in progress.
 	if hasConfigMap && hasSecret {
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("release %#q helmV3 migration in progress", key.ReleaseName(cr)))
+
+		found, err := findMigrationApp(ctx, cc.Clients.Helm, tillerNamespace)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if !found {
```
Comment (tomahawk28) [inline service/controller/app/resource/releasemigration/create.go:124]: If the operator found out remaining helm V2 releases from clusters, it should re-initiate installing the migration app.

File: service/controller/app/resource/releasemigration/create.go [hunk 2]
```diff
@@ -262,34 +277,93 @@ func (r *Resource) uncordonChart(ctx context.Context, g8sClient versioned.Interf
...
-func (r *Resource) hasHelmV2ConfigMaps(k8sClient kubernetes.Interface, releaseName, tillerNamespace string) (bool, error) {
+func (r *Resource) hasHelmV2ConfigMaps(k8sClient k8sclient.Interface, tillerNamespace string) (bool, error) {
+	chartMap, err := getChartMap(k8sClient, r.chartNamespace)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
 	lo := metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("%s=%s,%s=%s", "NAME", releaseName, "OWNER", "TILLER"),
+		LabelSelector: fmt.Sprintf("%s=%s", "OWNER", "TILLER"),
```
Comment (tomahawk28) [inline service/controller/app/resource/releasemigration/create.go:287]: We need to get the whole helm V2 configMaps, not just chart-operator.

File: service/controller/app/resource/releasemigration/create.go [hunk 3]
```diff
@@ -262,34 +277,93 @@ func (r *Resource) uncordonChart(ctx context.Context, g8sClient versioned.Interf
...
 		return false, microerror.Mask(err)
 	}
 
-	return len(cms.Items) > 0, nil
+	var count int
+	for _, cm := range cms.Items {
+		if _, ok := chartMap[cm.GetLabels()["NAME"]]; !ok {
+			continue
+		}
+		count++
```
Comment (tomahawk28) [inline service/controller/app/resource/releasemigration/create.go:301]: Counting helm releases only have relevant `charts` CRs.

File: service/controller/app/resource/releasemigration/create.go [hunk 4]
```diff
@@ -262,34 +277,93 @@ func (r *Resource) uncordonChart(ctx context.Context, g8sClient versioned.Interf
...
+			return false, microerror.Mask(err)
+		}
+
+		namespaces := map[string]bool{}
+		for _, chart := range list.Items {
+			namespaces[chart.Spec.Namespace] = true
+		}
+
+		for name, _ := range namespaces {
+			releaseNamespaces = append(releaseNamespaces, name)
```
Comment (tomahawk28) [inline service/controller/app/resource/releasemigration/create.go:321]: Helm V3 secrets are distributed on whole namespaces, we need to get the secrets from the namespace `charts` CRs specified.

File: service/controller/app/resource/releasemigration/resource.go [hunk 5]
```diff
@@ -138,11 +139,17 @@ func (r *Resource) ensureReleasesMigrated(ctx context.Context, k8sClient k8sclie
...
 	// Wait until all helm v2 release are deleted
 	o := func() error {
-		releases, err := r.findHelmV2Releases(k8sClient, tillerNamespace)
+		completed, err := checkMigrationJobStatus(k8sClient, "giantswarm")
 		if err != nil {
 			return microerror.Mask(err)
 		}
-		if len(releases) > 0 {
+
+		if !completed {
```
Comment (tomahawk28) [inline service/controller/app/resource/releasemigration/resource.go:147]: Checking migration jobs succeed in here.

File: service/controller/app/resource/releasemigration/resource.go [hunk 6]
```diff
@@ -157,7 +164,7 @@ func (r *Resource) ensureReleasesMigrated(ctx context.Context, k8sClient k8sclie
 		r.logger.LogCtx(ctx, "level", "debug", "message", "migration not complete")
 	}
 
-	b := backoff.NewConstant(5*time.Minute, 10*time.Second)
+	b := backoff.NewConstant(20*time.Minute, 10*time.Second)
```
Comment (tomahawk28) [inline service/controller/app/resource/releasemigration/resource.go:167]: Extending to 20 minutes to wait for migration completed.

**Comment by tomahawk28** [comment]
Tested in `gorgoth`

File: service/controller/app/resource/releasemigration/create.go [hunk 7]
```diff
@@ -115,6 +115,21 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
 	if hasConfigMap && hasSecret {
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("release %#q helmV3 migration in progress", key.ReleaseName(cr)))
+
+		found, err := findMigrationApp(ctx, cc.Clients.Helm, tillerNamespace)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if !found {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("release %#q had been purged during migration, reinstalling..", migrationApp))
```
Comment (rossf7) [inline service/controller/app/resource/releasemigration/create.go:125]:
```suggestion
			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("release %#q had been purged during migration, reinstalling...", migrationApp))
```

File: service/controller/app/resource/releasemigration/create.go [hunk 8]
```diff
@@ -262,34 +277,93 @@ func (r *Resource) uncordonChart(ctx context.Context, g8sClient versioned.Interf
...
+		return false, microerror.Mask(err)
+	}
+
+	return job.Status.Succeeded > 0, nil
+}
+
+func getChartMap(k8sClient k8sclient.Interface, namespace string) (map[string]bool, error) {
+	charts := make(map[string]bool)
+
+	// Get list of chart CRs as not all helm 2 releases will have a chart CR.
```
Comment (rossf7) [inline service/controller/app/resource/releasemigration/create.go:358]: ❤️  this comment. Important point to mention.

File: service/controller/app/resource/releasemigration/resource.go [hunk 9]
```diff
@@ -138,11 +139,17 @@ func (r *Resource) ensureReleasesMigrated(ctx context.Context, k8sClient k8sclie
 
 	// Wait until all helm v2 release are deleted
 	o := func() error {
-		releases, err := r.findHelmV2Releases(k8sClient, tillerNamespace)
+		completed, err := checkMigrationJobStatus(k8sClient, "giantswarm")
```
Comment (rossf7) [inline service/controller/app/resource/releasemigration/resource.go:142]: Nice better to use giantswarm.

**Pull request merged**